File: FlowMuse-App/lib/features/whiteboard/smart_layout/recognition/composition_hints.dart
part of 'recognition_models.dart';

/// 同轮结构扩展。只引用 unitId（进入语义文档后就是 blockId），不拥有正文。
class RecognitionCompositionHints {
  const RecognitionCompositionHints({
    this.pageIntent = 'unknown',
    this.sections = const [],
    this.mediaGroups = const [],
    this.softLineBreaks = const [],
  });

  static const version = 'composition-hints/1';
  final String pageIntent;
  final List<RecognitionSectionHint> sections;
  final List<RecognitionMediaGroup> mediaGroups;
  final List<RecognitionSoftBreak> softLineBreaks;

  factory RecognitionCompositionHints.fromJson(Object? value) {
    const r = RecognitionJsonReader(RecognitionParseSide.response);
    final m = r.object(value, 'compositionHints', const {
      'version',
      'pageIntent',
      'sections',
      'mediaGroups',
      'softLineBreaks',
    });
    if (m['version'] != version ||
        !const {
          'reading',
          'comparison',
          'mixed',
          'unknown',
        }.contains(m['pageIntent'])) {
      r.invalid('compositionHints', '版本或页面意图非法');
    }
    String id(Map<String, Object?> m, String key) {
      final v = r.string(m, key, 'compositionHints');
      if (v.trim().isEmpty || v.runes.length > 64) r.invalid(key, 'ID非法');
      return v;
    }

    List<String> ids(Map<String, Object?> m, String key) {
      final raw = r.list(m, key, 'compositionHints');
      if (raw.isEmpty ||
          raw.length > 128 ||
          raw.any(
            (v) => v is! String || v.trim().isEmpty || v.runes.length > 64,
          )) {
        r.invalid(key, '成员必须为非空ID数组且不超过128');
      }
      final out = raw.cast<String>();
      r.unique(out, key, '成员');
      return List.unmodifiable(out);
    }

    List<Object?> entries(String key) {
      final list = r.list(m, key, 'compositionHints');
      if (list.length > 128) r.invalid(key, '超过128');
      return list;
    }

    final sections = <RecognitionSectionHint>[];
    for (final raw in entries('sections')) {
      final s = r.object(raw, 'sections', const {
        'sectionId',
        'headingUnitId',
        'memberUnitIds',
      });
      sections.add(
        RecognitionSectionHint(
          sectionId: id(s, 'sectionId'),
          headingUnitId: id(s, 'headingUnitId'),
          memberUnitIds: ids(s, 'memberUnitIds'),
        ),
      );
    }
    final media = <RecognitionMediaGroup>[];
    for (final raw in entries('mediaGroups')) {
      final g = r.object(raw, 'mediaGroups', const {
        'groupId',
        'figureUnitIds',
        'textUnitIds',
        'confidence',
      });
      media.add(
        RecognitionMediaGroup(
          groupId: id(g, 'groupId'),
          figureUnitIds: ids(g, 'figureUnitIds'),
          textUnitIds: ids(g, 'textUnitIds'),
          confidence: r.unitInterval(g, 'confidence', 'mediaGroups'),
        ),
      );
    }
    final breaks = <RecognitionSoftBreak>[];
    for (final raw in entries('softLineBreaks')) {
      final b = r.object(raw, 'softLineBreaks', const {
        'unitId',
        'newlineIndexes',
        'confidence',
      });
      final indexes = r.list(b, 'newlineIndexes', 'softLineBreaks');
      var last = -1;
      if (indexes.isEmpty || indexes.length > 2000) {
        r.invalid('newlineIndexes', '换行编号为空或过多');
      }
      for (final index in indexes) {
        if (index is! int || index <= last) {
          r.invalid('newlineIndexes', '编号必须递增且非负');
        }
        last = index;
      }
      breaks.add(
        RecognitionSoftBreak(
          unitId: id(b, 'unitId'),
          newlineIndexes: List.unmodifiable(indexes.cast<int>()),
          confidence: r.unitInterval(b, 'confidence', 'softLineBreaks'),
        ),
      );
    }
    r.unique(sections.map((s) => s.sectionId), 'sections', 'sectionId');
    r.unique(media.map((g) => g.groupId), 'mediaGroups', 'groupId');
    r.unique(breaks.map((b) => b.unitId), 'softLineBreaks', 'unitId');
    return RecognitionCompositionHints(
      pageIntent: m['pageIntent']! as String,
      sections: List.unmodifiable(sections),
      mediaGroups: List.unmodifiable(media),
      softLineBreaks: List.unmodifiable(breaks),
    );
  }

  /// 网络与纠错共用关系校验；无请求上下文时仍检查角色/成员/顺序。
  void validate({
    List<RecognitionUnitInput>? units,
    required List<String> readingOrder,
    required Map<String, String> roles,
    required List<RecognitionListGroup> listGroups,
    required List<RecognitionCaption> captions,
    required bool hasOverview,
  }) {
    const r = RecognitionJsonReader(RecognitionParseSide.response);
    Never fail() => r.invalid('compositionHints', '角色、成员或连续性无效');
    final byId = {
      for (final u in units ?? <RecognitionUnitInput>[]) u.unitId: u,
    };
    final positions = {
      for (var i = 0; i < readingOrder.length; i++) readingOrder[i]: i,
    };
    final subtrees = <Set<String>>[];
    for (final root in listGroups) {
      final set = root.members.toSet();
      for (var i = 0; i < listGroups.length; i++) {
        for (final child in listGroups) {
          if (set.contains(child.parentUnitId)) set.addAll(child.members);
        }
      }
      if (set.any((id) => roles[id] != 'listItem')) fail();
      subtrees.add(set);
    }
    void continuous(List<String> ids) {
      final set = ids.toSet();
      if (ids.isEmpty ||
          ids.length != set.length ||
          ids.any((id) => !positions.containsKey(id))) {
        fail();
      }
      final indices = ids.map((id) => positions[id]!).toList()..sort();
      if (indices.last - indices.first + 1 != ids.length) fail();
      for (final subtree in subtrees) {
        final hits = set.intersection(subtree).length;
        if (hits != 0 && hits != subtree.length) fail();
      }
    }

    final sectionOf = <String, String>{};
    for (final s in sections) {
      if (roles[s.headingUnitId] != 'title' || s.memberUnitIds.isEmpty) fail();
      final ids = [s.headingUnitId, ...s.memberUnitIds];
      continuous(ids);
      var last = positions[s.headingUnitId]!;
      for (final id in s.memberUnitIds) {
        if (positions[id]! <= last || roles[id] == 'title') fail();
        last = positions[id]!;
      }
      for (final id in ids) {
        if (sectionOf.containsKey(id)) fail();
        sectionOf[id] = s.sectionId;
      }
    }
    final captionOf = <String, String>{};
    for (final c in captions) {
      if (roles[c.captionUnitId] != 'caption' ||
          !positions.containsKey(c.targetUnitId) ||
          captionOf.containsKey(c.captionUnitId) ||
          sectionOf[c.captionUnitId] != sectionOf[c.targetUnitId] ||
          (units != null &&
              !const {
                RecognitionUnitKind.figure,
                RecognitionUnitKind.preserved,
              }.contains(byId[c.targetUnitId]?.kind))) {
        fail();
      }
      captionOf[c.captionUnitId] = c.targetUnitId;
    }
    if (roles.entries.any(
      (e) => e.value == 'caption' && !captionOf.containsKey(e.key),
    )) {
      fail();
    }
    final owners = <String>{};
    for (final g in mediaGroups) {
      if (!hasOverview ||
          g.figureUnitIds.isEmpty ||
          g.textUnitIds.isEmpty ||
          !g.confidence.isFinite ||
          g.confidence < 0 ||
          g.confidence > 1 ||
          g.figureUnitIds.any(
            (id) => units != null
                ? byId[id]?.kind != RecognitionUnitKind.figure
                : roles.containsKey(id),
          ) ||
          g.textUnitIds.any(
            (id) => roles[id] != 'body' && roles[id] != 'listItem',
          )) {
        fail();
      }
      final ids = [
        ...g.figureUnitIds,
        ...g.textUnitIds,
        for (final c in captions)
          if (g.figureUnitIds.contains(c.targetUnitId)) c.captionUnitId,
      ];
      continuous(ids);
      for (final id in ids) {
        if (!owners.add(id) || sectionOf[id] != sectionOf[ids.first]) fail();
      }
    }
    for (final target in captionOf.values.toSet()) {
      if (owners.contains(target)) continue;
      continuous([
        target,
        for (final c in captions)
          if (c.targetUnitId == target) c.captionUnitId,
      ]);
    }
    for (final b in softLineBreaks) {
      if (!const {'title', 'body', 'caption'}.contains(roles[b.unitId])) fail();
      if (units == null) continue;
      final u = byId[b.unitId];
      if (u?.kind != RecognitionUnitKind.ink || u?.text == null) fail();
      final lines = u!.text!.split('\n');
      for (final index in b.newlineIndexes) {
        if (index < 0 ||
            index + 1 >= lines.length ||
            lines[index].trim().isEmpty ||
            lines[index + 1].trim().isEmpty) {
          fail();
        }
      }
    }
  }

  Map<String, Object?> toJson() => {
    'version': version,
    'pageIntent': pageIntent,
    'sections': [for (final s in sections) s.toJson()],
    'mediaGroups': [for (final g in mediaGroups) g.toJson()],
    'softLineBreaks': [for (final b in softLineBreaks) b.toJson()],
  };

  @override
  bool operator ==(Object other) =>
      other is RecognitionCompositionHints &&
      pageIntent == other.pageIntent &&
      _listEq(sections, other.sections) &&
      _listEq(mediaGroups, other.mediaGroups) &&
      _listEq(softLineBreaks, other.softLineBreaks);
  @override
  int get hashCode => Object.hash(
    pageIntent,
    sections.length,
    mediaGroups.length,
    softLineBreaks.length,
  );
}

class RecognitionSectionHint {
  const RecognitionSectionHint({
    required this.sectionId,
    required this.headingUnitId,
    required this.memberUnitIds,
  });
  final String sectionId;
  final String headingUnitId;
  final List<String> memberUnitIds;
  Map<String, Object?> toJson() => {
    'sectionId': sectionId,
    'headingUnitId': headingUnitId,
    'memberUnitIds': [...memberUnitIds],
  };
  @override
  bool operator ==(Object other) =>
      other is RecognitionSectionHint &&
      sectionId == other.sectionId &&
      headingUnitId == other.headingUnitId &&
      _listEq(memberUnitIds, other.memberUnitIds);
  @override
  int get hashCode =>
      Object.hash(sectionId, headingUnitId, memberUnitIds.length);
}

class RecognitionMediaGroup {
  const RecognitionMediaGroup({
    required this.groupId,
    required this.figureUnitIds,
    required this.textUnitIds,
    required this.confidence,
  });
  final String groupId;
  final List<String> figureUnitIds;
  final List<String> textUnitIds;
  final double confidence;
  Map<String, Object?> toJson() => {
    'groupId': groupId,
    'figureUnitIds': [...figureUnitIds],
    'textUnitIds': [...textUnitIds],
    'confidence': confidence,
  };
  @override
  bool operator ==(Object other) =>
      other is RecognitionMediaGroup &&
      groupId == other.groupId &&
      confidence == other.confidence &&
      _listEq(figureUnitIds, other.figureUnitIds) &&
      _listEq(textUnitIds, other.textUnitIds);
  @override
  int get hashCode => Object.hash(groupId, confidence);
}

class RecognitionSoftBreak {
  const RecognitionSoftBreak({
    required this.unitId,
    required this.newlineIndexes,
    required this.confidence,
  });
  final String unitId;
  final List<int> newlineIndexes;
  final double confidence;
  Map<String, Object?> toJson() => {
    'unitId': unitId,
    'newlineIndexes': [...newlineIndexes],
    'confidence': confidence,
  };
  @override
  bool operator ==(Object other) =>
      other is RecognitionSoftBreak &&
      unitId == other.unitId &&
      confidence == other.confidence &&
      _listEq(newlineIndexes, other.newlineIndexes);
  @override
  int get hashCode => Object.hash(unitId, confidence);
}

File: FlowMuse-Server/internal/layoutrecognitionv3/composition_hints.go
package layoutrecognitionv3

import (
	"strings"
	"unicode/utf8"
)

// CompositionHints 只引用本轮 unit，不拥有正文、坐标或源账本。
type CompositionHints struct {
	Version        string           `json:"version"`
	PageIntent     string           `json:"pageIntent"`
	Sections       []SectionHint    `json:"sections"`
	MediaGroups    []MediaGroupHint `json:"mediaGroups"`
	SoftLineBreaks []SoftBreakHint  `json:"softLineBreaks"`
}

type SectionHint struct {
	SectionID     string   `json:"sectionId"`
	HeadingUnitID string   `json:"headingUnitId"`
	MemberUnitIDs []string `json:"memberUnitIds"`
}

type MediaGroupHint struct {
	GroupID       string   `json:"groupId"`
	FigureUnitIDs []string `json:"figureUnitIds"`
	TextUnitIDs   []string `json:"textUnitIds"`
	Confidence    *float64 `json:"confidence"`
}

type SoftBreakHint struct {
	UnitID         string   `json:"unitId"`
	NewlineIndexes []int    `json:"newlineIndexes"`
	Confidence     *float64 `json:"confidence"`
}

// 即便值为空串/null，结构输出也不得携带正文；只在新协商模式启用。
func containsStructureText(value any) bool {
	switch node := value.(type) {
	case map[string]any:
		for key, child := range node {
			if key == "text" || containsStructureText(child) {
				return true
			}
		}
	case []any:
		for _, child := range node {
			if containsStructureText(child) {
				return true
			}
		}
	}
	return false
}

func validateCompositionHints(req *RecognitionRequest, result *ModelStructureResult) *WireError {
	fail := func() *WireError {
		return wireErr(CodeInvalidProviderResp, "composition 提示的字段、角色、成员或连续性无效")
	}
	h := result.CompositionHints
	if h == nil || h.Version != "composition-hints/1" ||
		h.Sections == nil || h.MediaGroups == nil || h.SoftLineBreaks == nil ||
		result.ReadingOrder == nil || result.Roles == nil || result.ListGroups == nil || result.Captions == nil || result.Warnings == nil {
		return fail()
	}
	if h.PageIntent != "reading" && h.PageIntent != "comparison" && h.PageIntent != "mixed" && h.PageIntent != "unknown" {
		return fail()
	}
	if len(h.Sections) > len(req.Units) || len(h.MediaGroups) > len(req.Units) || len(h.SoftLineBreaks) > len(req.Units) {
		return fail()
	}
	units := map[string]UnitInput{}
	roles := map[string]string{}
	positions := map[string]int{}
	for _, u := range req.Units {
		units[u.UnitID] = u
	}
	for _, r := range result.Roles {
		if r.Text != nil {
			return fail()
		}
		roles[r.UnitID] = r.Role
	}
	for i, id := range result.ReadingOrder {
		positions[id] = i
	}
	validID := func(id string) bool { return strings.TrimSpace(id) != "" && utf8.RuneCountInString(id) <= 64 }
	// 每个列表完整子树是不可拆单元，而不是仅检查各级成员。
	subtrees := []map[string]bool{}
	for _, root := range result.ListGroups {
		if root.Text != nil {
			return fail()
		}
		set := map[string]bool{}
		for _, id := range root.Members {
			set[id] = true
		}
		for range result.ListGroups {
			for _, child := range result.ListGroups {
				if child.ParentUnitID != nil && set[*child.ParentUnitID] {
					for _, id := range child.Members {
						set[id] = true
					}
				}
			}
		}
		for id := range set {
			if roles[id] != "listItem" {
				return fail()
			}
		}
		subtrees = append(subtrees, set)
	}
	continuous := func(ids []string) bool {
		if len(ids) == 0 {
			return false
		}
		set, low, high := map[string]bool{}, len(positions), -1
		for _, id := range ids {
			p, ok := positions[id]
			if !ok || set[id] {
				return false
			}
			set[id] = true
			low, high = min(low, p), max(high, p)
		}
		if high-low+1 != len(set) {
			return false
		}
		for _, subtree := range subtrees {
			hits := 0
			for id := range subtree {
				if set[id] {
					hits++
				}
			}
			if hits != 0 && hits != len(subtree) {
				return false
			}
		}
		return true
	}
	sectionOf, sectionIDs := map[string]string{}, map[string]bool{}
	for _, s := range h.Sections {
		if !validID(s.SectionID) || sectionIDs[s.SectionID] || roles[s.HeadingUnitID] != "title" || len(s.MemberUnitIDs) == 0 {
			return fail()
		}
		sectionIDs[s.SectionID] = true
		ids := append([]string{s.HeadingUnitID}, s.MemberUnitIDs...)
		if !continuous(ids) {
			return fail()
		}
		last := positions[s.HeadingUnitID]
		for _, id := range s.MemberUnitIDs {
			if positions[id] <= last || roles[id] == "title" {
				return fail()
			}
			last = positions[id]
		}
		for _, id := range ids {
			if sectionOf[id] != "" {
				return fail()
			}
			sectionOf[id] = s.SectionID
		}
	}
	captionOf := map[string]string{}
	for _, c := range result.Captions {
		target, ok := units[c.TargetUnitID]
		if c.Text != nil || roles[c.CaptionUnitID] != "caption" || !ok ||
			(target.Kind != "figure" && target.Kind != "preserved") || captionOf[c.CaptionUnitID] != "" ||
			sectionOf[c.CaptionUnitID] != sectionOf[c.TargetUnitID] {
			return fail()
		}
		captionOf[c.CaptionUnitID] = c.TargetUnitID
	}
	for id, role := range roles {
		if role == "caption" && captionOf[id] == "" {
			return fail()
		}
	}
	owner, groupIDs := map[string]string{}, map[string]bool{}
	for _, g := range h.MediaGroups {
		if req.OverviewPngBase64 == "" || !validID(g.GroupID) || groupIDs[g.GroupID] || len(g.FigureUnitIDs) == 0 || len(g.TextUnitIDs) == 0 ||
			g.Confidence == nil || !(*g.Confidence >= 0 && *g.Confidence <= 1) {
			return fail()
		}
		groupIDs[g.GroupID] = true
		ids := append(append([]string{}, g.FigureUnitIDs...), g.TextUnitIDs...)
		figures := map[string]bool{}
		for _, id := range g.FigureUnitIDs {
			if units[id].Kind != "figure" {
				return fail()
			}
			figures[id] = true
		}
		for _, id := range g.TextUnitIDs {
			if roles[id] != "body" && roles[id] != "listItem" {
				return fail()
			}
		}
		for caption, target := range captionOf {
			if figures[target] {
				ids = append(ids, caption)
			}
		}
		if !continuous(ids) {
			return fail()
		}
		for _, id := range ids {
			if owner[id] != "" || sectionOf[id] != sectionOf[ids[0]] {
				return fail()
			}
			owner[id] = g.GroupID
		}
	}
	// 没有正文 mediaGroup 的单图+图注组件也必须连续。
	for _, c := range result.Captions {
		if owner[c.TargetUnitID] != "" {
			continue
		}
		ids := []string{c.TargetUnitID}
		for caption, target := range captionOf {
			if target == c.TargetUnitID {
				ids = append(ids, caption)
			}
		}
		if !continuous(ids) {
			return fail()
		}
	}
	seenBreaks := map[string]bool{}
	for _, b := range h.SoftLineBreaks {
		u, ok := units[b.UnitID]
		role := roles[b.UnitID]
		if !ok || u.Kind != "ink" || u.Text == nil || (role != "title" && role != "body" && role != "caption") ||
			seenBreaks[b.UnitID] || len(b.NewlineIndexes) == 0 || b.Confidence == nil || !(*b.Confidence >= 0 && *b.Confidence <= 1) {
			return fail()
		}
		seenBreaks[b.UnitID] = true
		lines := strings.Split(*u.Text, "\n")
		last := -1
		for _, index := range b.NewlineIndexes {
			if index <= last || index+1 >= len(lines) || strings.TrimSpace(lines[index]) == "" || strings.TrimSpace(lines[index+1]) == "" {
				return fail()
			}
			last = index
		}
	}
	return nil
}

const compositionHintPrompt = `
本次启用整页内容组织，在同一 JSON 增加 compositionHints，禁止返回 figureTextLinks：
"compositionHints":{"version":"composition-hints/1","pageIntent":"reading|comparison|mixed|unknown","sections":[{"sectionId":"s1","headingUnitId":"标题ID","memberUnitIds":["有序成员ID"]}],"mediaGroups":[{"groupId":"m1","figureUnitIds":["图片ID"],"textUnitIds":["正文ID"],"confidence":0.95}],"softLineBreaks":[{"unitId":"OCR单元ID","newlineIndexes":[0],"confidence":0.95}]}
- 所有数组必须存在，无证据用 []，不用 null；只返回引用/编号/置信度，不得夹带正文或坐标。
- pageIntent 表示阅读、同级比较或混合。sections 仅一层章节，heading 必须 title，成员不含自己的 heading，章节连续且不交叉；页总标题可在章节之外。
- 根据概览中图片的实际对象及全文语义判断图文对应，不按最近距离硬配。mediaGroup 至少一图和一段 body/listItem；可以多图共用一段说明，该文字只出现一次；一单元最多属于一组。不确定不配。
- 短图注保留 caption 角色及 captions 关系，不列入 mediaGroups.textUnitIds；长解释仍为 body/listItem。每个 caption 必须挂靠 figure/preserved。组的图片、正文及其图注在 readingOrder 中连续，不夹无关单元，不跨章节、不拆完整列表子树；不要重写正文来满足约束。
- softLineBreaks 只标记 ink 中因手写行宽产生的物理换行（从第一个 \n 编号 0 开始，递增且不重复）。明确为普通正文、标题或短图注，且合并后语义仍为同一句时才给高置信。原生 typed、空行分段、列表、代码、公式、诗歌、未知内容不得合并；不改拼写/标点/连字符，不返回改写文本。
- 不确定的换行或关联留空。原 readingOrder/roles/listGroups/captions/warnings 仍必须完整返回。`
